plugins/goodix-tp/fu-goodixtp-brlb.c

```diff
@@ -0,0 +1,519 @@
+		fw_summary->subsys[i].size = *(guint32 *)&data[info_offset + 1];
+		fw_summary->subsys[i].flash_addr = *(guint32 *)&data[info_offset + 5];
+		if ((gint)fw_offset > (gint)firmware_size) {
+			g_debug("Sybsys offset exceed Firmware size");
+			return FALSE;
+		}
+		fw_summary->subsys[i].data = data + fw_offset;
+		fw_offset += fw_summary->subsys[i].size;
+	}
+
+	memcpy(tmp_buf, fw_summary->fw_pid, 8);
+	g_debug("Firmware package protocol: V%u", fw_summary->protocol_ver);
+	g_debug("Firmware PID:GT%s", tmp_buf);
+	g_debug("Firmware VID:%02x %02x %02x %02x",
+		fw_summary->fw_vid[0],
+		fw_summary->fw_vid[1],
+		fw_summary->fw_vid[2],
+		fw_summary->fw_vid[3]);
+	g_debug("Firmware chip type:%02X", fw_summary->chip_type);
+	g_debug("Firmware size:%u", fw_summary->size);
+	g_debug("Firmware subsystem num:%u", fw_summary->subsys_num);
+
+	if (brlb_update_ctrl.has_config) {
+		cfg_ver = data[firmware_size + 64 + 34];
+		g_debug("cfg_ver:%02x", cfg_ver);
+	}
+
+	brlb_update_ctrl.bin_version =
+	    (fw_summary->fw_vid[2] << 16) | (fw_summary->fw_vid[3] << 8) | cfg_ver;
+
+	return TRUE;
+}
+
+static gboolean
+brlb_update_prepare(FuDevice *device, GError **error)
+{
+	guint8 tempBuf[5] = {0};
+	guint8 recvBuf[5] = {0};
+	gint retry = 3;
+	gboolean ret;
+
+	/* step 1. switch mini system */
+	tempBuf[0] = 0x01;
+	ret = send_cmd(device, 0x10, tempBuf, 1, error);
+	if (!ret) {
+		g_debug("Failed send minisystem cmd");
+		return ret;
+	}
+	while (retry--) {
+		fu_device_sleep(device, 200);
+		ret = hid_read(device, 0x10010, tempBuf, 1, error);
+		if (ret && tempBuf[0] == 0xDD)
+			break;
+	}
+	if (retry < 0) {
+		g_debug("Failed switch minisystem flag=0x%02x", tempBuf[0]);
+		return FALSE;
+	}
+	g_debug("Switch mini system successfully");
+
+	/* step 2. erase flash */
+	tempBuf[0] = 0x01;
+	ret = send_cmd(device, 0x11, tempBuf, 1, error);
+	if (!ret) {
+		g_debug("Failed send erase flash cmd");
+		return ret;
+	}
+
+	retry = 10;
+	memset(tempBuf, 0x55, 5);
+	while (retry--) {
+		fu_device_sleep(device, 10);
+		ret = hid_write(device, 0x14000, tempBuf, 5, error);
+		if (!ret) {
+			g_debug("Failed write sram");
+			return ret;
+		}
+		ret = hid_read(device, 0x14000, recvBuf, 5, error);
+		if (!memcmp(tempBuf, recvBuf, 5))
+			break;
+	}
+	if (retry < 0) {
+		g_debug("Read back failed, buf:%02x %02x %02x %02x %02x\n",
+			recvBuf[0],
+			recvBuf[1],
+			recvBuf[2],
+			recvBuf[3],
+			recvBuf[4]);
+		return FALSE;
+	}
+
+	g_debug("Updata prepare OK");
+	return TRUE;
+}
+
+static gboolean
+flash_sub_system(FuDevice *device, struct fw_subsys_info *subsys, GError **error)
+{
+	guint32 data_size = 0;
+	guint32 offset = 0;
+	guint32 temp_addr = subsys->flash_addr;
+	guint32 total_size = subsys->size;
+	guint32 checksum;
+	guint8 cmdBuf[10] = {0};
+	guint8 flag;
+	gint resend_rty = 3;
+	gint retry;
+	guint32 i;
+	gboolean ret;
+
+	while (total_size > 0) {
+		data_size = total_size > 4096 ? 4096 : total_size;
+	resend:
```

Comment:
please use fu_device_retry instead -- that will make sure it works properly in all cases.